daemon/containerd/image_push.go

```diff
@@ -2,15 +2,255 @@ package containerd
 
 import (
 	"context"
-	"errors"
+	"fmt"
 	"io"
+	"strings"
 
+	"github.com/containerd/containerd/content"
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/images"
+	containerdimages "github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/platforms"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
 	"github.com/docker/distribution/reference"
 	"github.com/docker/docker/api/types/registry"
 	"github.com/docker/docker/errdefs"
+	"github.com/docker/docker/pkg/streamformatter"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/semaphore"
 )
 
-// PushImage initiates a push operation on the repository named localName.
-func (i *ImageService) PushImage(ctx context.Context, ref reference.NamedTagged, metaHeaders map[string][]string, authConfig *registry.AuthConfig, outStream io.Writer) (outerr error) {
-	return errdefs.NotImplemented(errors.New("not implemented"))
+// PushImage initiates a push operation of the image pointed to by targetRef.
+// Image manifest (or index) is pushed as is, which will probably fail if you
+// don't have all content referenced by the index.
+// Cross-repo mounts will be attempted for non-existing blobs.
+//
+// It will also add distribution source labels to the pushed content
+// pointing to the new target repository. This will allow subsequent pushes
+// to perform cross-repo mounts of the shared content when pushing to a different
+// repository on the same registry.
+func (i *ImageService) PushImage(ctx context.Context, targetRef reference.Named, metaHeaders map[string][]string, authConfig *registry.AuthConfig, outStream io.Writer) error {
+	if _, tagged := targetRef.(reference.Tagged); !tagged {
+		if _, digested := targetRef.(reference.Digested); !digested {
+			return errdefs.NotImplemented(errors.New("push all tags is not implemented"))
+		}
+	}
+
+	leasedCtx, release, err := i.client.WithLease(ctx)
+	if err != nil {
+		return err
+	}
+	defer release(leasedCtx)
```

Comment:
Technically, if it returned an error then the lease couldn't be deleted, so that might be something worthy of logging. However, looking at how containerd itself uses this, it's "common" (I haven't really seen any code that does it differently) to just call it and ignore the error.
Looking at the code, the only case where this could happen if the lease deletion fails. This can happen if the lease no longer exists (which is already the state we want) or the c8d connection is broken, which will pop out in some other place anyway.

Reply:
Gotcha; yeah, TBH, containerd ignoring things could _sometimes_ be a guideline, but we should definitely check ourselves what makes sense (mistakes are made by everyone, so it's always possible containerd is on the wrong on these, in which case we could contribute there as well).

- "not found" errors; yup definitely we should ignore
- more curious; could "something else using the files / mounts / whatever" cause things to not be cleaned up? (in which case we may want to have a log to backtrace we potentially leaked things somewhere); of course that is if that wouldn't trigger "false positives" (race conditions, things being "eventually" cleaned up etc).

Reply:
This only deletes a lease - which is created only in this call-site and deleting it shouldn't fail unless the lease was already deleted, containerd connection is broken or containerd state is corrupted. In which case we will have loud errors in other places. Even if we fail to delete the lease and containerd is fully operational with the lease state still being present, the lease will expire after 24h.
It doesn't harm to log it though.. Except it's just bloating a simple one line defer to a deferred anonymous function call with a conditional log invocation 😅  But maybe it can save us from some trouble if the Lease manager has some major future changes from the current state, so I'm not that much against logging it.

Reply:
Thanks! Yes I didn't know if actual _mounts_ and/or _files_ were involved here. I'm mostly "fine" either way I guess